src/dmd/access.d

```diff
@@ -612,8 +612,24 @@ public Dsymbol mostVisibleOverload(Dsymbol s)
         else
             break;
 
-        if (next && mostVisible.prot().isMoreRestrictiveThan(next.prot()))
+        if (next && mostVisible.effectiveProtAt(mod).isMoreRestrictiveThan(next.effectiveProtAt(mod)))
             mostVisible = next;
     }
     return mostVisible;
 }
+
+/**
+ * Return the "effective" protection attribute of a symbol seen from a module.
+ * The effective protection attribute is the same as the regular protection attribute,
+ * except package() is "private" if the module is outside the package;
+ * otherwise, "public".
+ */
+Prot effectiveProtAt(Dsymbol d, Module mod)
```

Comment:
I also recommend making this a nested function in the `mostVisibleOverload` function, or at a minimum, make it `private`.

Reply:
Done, and renamed to protectionIn